worlds/witness/__init__.py

```diff
@@ -78,45 +83,60 @@ def generate_early(self):
                 raise Exception("This Witness world doesn't have any progression items. Please turn on Symbol Shuffle,"
                                 " Door Shuffle or Laser Shuffle.")
 
-        self.player_logic = WitnessPlayerLogic(self.multiworld, self.player)
+    def create_regions(self):
+        """
+        Doing all of this in create_regions because I want to have access to pre-collected items,
+        which influence rules for locations as well as region connections.
+        (Pre-collected items are not available in generate_early, but generate_basic would be too late,
```

Comment:
precollected items are available as multiworld.start_inventory[player]: dict[str, int] during generate_early, which then gets turned into precollectedd items on state between generate_early and create_regions

Reply:
Fair enough. alwaysintreble's comments made me feel like I wasn't supposed to use that, but rereading them, they didn't actually technically say that

Reply:
Ok, so it seems to be working. One thing that was confusing is that multiworld.start_inventory[player] is *not* a dict[str,int], multiworld.start_inventory[player].value is. But I worked that out after a bit of trial and error

Reply:
either should work fine. the generate_early approach should be more performant as it doesn't assemble a state that is then messed with again.